src/rust_connection/fd_read_write.rs

```diff
@@ -235,17 +219,33 @@ impl<W: WriteFD> WriteFD for BufWriteFD<W> {
         }
     }
 
-    fn flush(&mut self) -> Result<()> {
-        self.flush_buffer().and_then(|_| self.inner.flush())
+    fn flush(&mut self, block: bool) -> Result<()> {
+        self.flush_buffer(block)
+            .and_then(|_| self.inner.flush(block))
     }
 }
 
-/// A version of [`std::io::Read`] that also allows receiving file descriptors.
-pub trait ReadFD {
-    /// Read some bytes and FDs from this reader, returning how many bytes were read.
+impl<T: WriteFD> Poll for BufWriteFD<T> {
+    fn poll(&mut self, read: bool, write: bool) -> Result<(bool, bool)> {
+        if write && self.data_buf.len() != self.data_buf.capacity() {
```

Comment:
```suggestion
        if write && !self.data_buf.is_empty() {
```
(Your version checks if the write buffer is full, not empty)

Reply:
As long as it isn't full, we can write to it, can't we? I think its the comment below what needs fixing.

Reply:
Well... yes and no. Depends on the exact semantics we want.
Due to this, any kind of write can always fail with `WouldBlock`, no matter how small:
https://github.com/psychon/x11rb/blob/master/src/rust_connection/fd_read_write.rs#L210-L211

So, if the semantics is supposed to be "we can write at least one byte", then... dunno, something is wrong.

But you are right, I was totally confused with my suggestion above.

Hm... For what we need, it might be okay for `BufWriteFD` to always defer to `inner`. If `inner` is writable, then so is `BufWriteFD`. If `inner` is not writeable, then... well, we cannot make any guarantees again. `BufWriteFD` tries to flush the buffer first and does not accept new data until the old buffer is fully flushed. So this is neither better nor worse than what your commit implements.

I'll let you decide what's best here. Everything seems to be "half weird".

Reply:
> https://github.com/psychon/x11rb/blob/master/src/rust_connection/fd_read_write.rs#L210-L211

Oh, I didn't see that. Let me think something...

Reply:
Fixed in https://github.com/psychon/x11rb/pull/426/commits/6c33745821e6eeb9307e1088b90e888d876574da. `write(_vectored)` will not return `WouldBlock` if there is any buffer available.